### ISSUE TITLE
Experiment: a better code intel panel

### ISSUE THREAD

**Issue by mrnugget** [sourcegraph/sourcegraph#30227]
### Goal

Create a better code intelligence panel to display definition, references, implemenations.

Turn the reference panel into a serious tool that doesn't make users fall back to their editor/IDE.

### Problems with current panel

* There are some general UX problems with the panel due to how it is integrated in our Sourcegraph web app and how it loads data: https://github.com/sourcegraph/sourcegraph/issues/30022
* It doesn't stay open when navigating through references: https://github.com/sourcegraph/sourcegraph/issues/13722
* It's not possible to filter references: https://github.com/sourcegraph/sourcegraph/issues/28480
* It's not possible to exclude file types 
* Navigating through references is incredibly slow
* It's not possible to scroll through full list of references quickly (you need to load, scroll, load some more, scroll, load some more, scroll, load some more)

### Plan

* **Stage 1**: Experiment to see what's possible - don't worry about merging, don't worry about `code-intel-extensions`
  * Stays open (maybe use side-panel to show references)
  * Loads fast
  * Keeps position in list of references 
  * Show how many references there are
  * Allows filtering (just simple `string.Contains(reference.filePath, userInput)`, which is by far enough as a first iteration)
  * Allow deleting entries from list (like VS Code does)
  * Maybe we don't need hover tooltip anymore if panel is really good (see reference panel at [cs.opensource.google](https://cs.opensource.google/go/go/+/master:src/bytes/buffer.go;l=30;drc=master))
* **Stage 2**: Build a feature-flagged **new panel** that others can play around with to get feedback
* **Stage 3**: **Maybe**: backport required features to extension API & `code-intel-extensions`
* **Stage 4**: Ship a better panel

**Comment by mrnugget** [comment]
Little update.

### Current status

https://user-images.githubusercontent.com/1185253/152131980-05f98df7-96df-493e-bc54-b61237471b66.mp4

Branch/PR: `mrn/cool-code-intel`/https://github.com/sourcegraph/sourcegraph/pull/30095 (see PR description for how to try this locally)

* New reference panel, added alongside our existing codebase, ~700 lines of code right now
* Reference panel stays open when navigating through references (would fix https://github.com/sourcegraph/sourcegraph/issues/13722)
* References are filterable (would fix https://github.com/sourcegraph/sourcegraph/issues/28480) 
* Definitions in same panel as references, no need to navigate to other panel
* Lists are collapsible
* Clicking on a reference opens little code blob on side so that main code blob doesn't navigate away
* Uses a single GraphQL query to fetch hover/definition/references

### What's missing

The following are **things that I want to fix before merging this** and making it available to teammates for getting first playing-around feedback:

* It's not behind a user-setting feature flag
* Some of the existing code is commented out/changed. It should instead be toggled on/off depending on feature flag.
* The buttons in the hover tooltips open the old panels.
  * **Idea**: remove them altogether for this experiment, or rewire them to open the panel.
* Panel can't be closed
* No filename shown above little code blob that's loaded inside reference panel
* Scrolling inside the reference panel and code blob loaded inside panel works only with a hack that I want to get rid of (I set a hardcoded `max-height` of both elements)
* Remove the "Token" panel that I only used for debugging

Then there are other **things missing that would make this better even as an experiment** (to accurately compare it to existing panel):
* Only LSIF data is used (no mixing of precise & search and no fallback to search)
* No pagination of references (only loads the first batch and doesn't show whether there's more data available)
* Filtering/globbing is really basic (`strings.Contains(reference.filePath, filterValue)`)
  * In order for pagination + filtering to work we'd need to do it in database layer
  * It doesn't support excluding files (which is a common usecase, people want to exclude all `*_test.go` files, for example)
* Better design
* Allow opening a reference code blob in "big blog" (like Google CS)
* Display a stack of definitions a user jumped along in refernce code blob (like Google CS), easy to do with this

And here's an **_incomplete_ list of things we'd need to think about and implement if we were to switch from existing panel to this**:
* Handle URL search query and parameters to be backwards compatible
* Handle URL in general: when should the URL be updated and when not, what can a user link to and what not
* Figure out whether to jump to definition on click and open panel when clicking on def (that's what Google CS does) or whether to open panel on click and then let user jump to def by clicking on that (current state)
* Figure out what to do with existing hover tooltip and its buttons.
  * Technically we don't need it anymore since we can show hover information in the panel instead on click
  * Means we need to think about click behaviour & hover tooltip. We could untie click & tooltip and use stencils instead and turn tokens into links on load. (Current implementation  of click2jump changes tokens into links when hover tooltip appears)
* Include the "History" panel  (or figure out what to do with the history panel)

### Next steps for me

Before end of iteration:

* I want to implement the "**things that I want to fix before merging this**" mentioned above
* Merge this behind feature flag, get others to play around with it

Next iteration:
* Write up RFC/doc to
  * examine problems (current codebase hard to extend due to everything having to go through extension API, unversioned extensions, etc.)
  * share learnings from this experiment (~7 days of work vs certainly multiples of that if I had to implement it via extenion, etc.)
  * sketch out what it would take to implement this in our extension
  * sketch out what it would take to implement this without extension (means: implementing the list of missing things above)
* Get input from growth-and-integrations-team, get some stats on extensions from analytics team
* make decision

**Comment by macraig** [comment]
The update looks amazing @mrnugget ! Can we loop in @jjinnii before merging and making this available to users?

**Comment by mrnugget** [comment]
Yeah, absolutely. But to be clear: making it available to users that are _not_ Sourcegraph employees is still a long way off. Want I want (and what I think I'll get) is a general  👍 / 👎  indicator from colleagues on whether they'd prefer this over the existing panel. Only then and once we've figured out a technical path forward (with or without extensions) would I think about making this available to more users.

**Comment by jjinnii** [comment]
Thanks @mrnugget ! Thanks for the update. It's looking really great, and I'd love to think through at this point how we incorporate design and how we can get to a state where we can get this in front of users with @macraig . 

I don't think think the RFC on extensions and the technical decision necessarily needs to happen before we start taking a look at design. Especially given it seems near ready to put behind a feature flag for internal feedback: 

- Should we do some recorded sessions/testing with internal users? 
- If we get enough feedback and they're mostly 👍, I'd love to move this prototype into design refinement. Refinement meaning taking a more systematic view of the design, and if we have the time, additional user testing. 

I know we're all really excited to ship this feature, but given that this exploratory work started out tackling a specific issue of keeping the ref panel open and became broader in scope, I think it's all the more reason to be more deliberate about how we move forward. 🙂

**Comment by mrnugget** [comment]
Yes, absolutely, @jjinnii! How would you prefer to incorporate design?

### The Plan

Here's how I see a possible plan:

* I can continue fixing papercuts, then ship prototype to for internal playing around with
* At the same time: you can start approaching this from the design perspective

Once we then go into the next stage we should be on the same page. Let me draw that up real quick...

<img width="674" alt="screenshot_2022-02-04_11 52 27@2x" src="https://user-images.githubusercontent.com/1185253/152516828-918d7a14-6662-4759-9068-c3f7c547a267.png">

Let me know if that matches what you had in mind :)

---

### Design Stuff To Chew On

Here are the things I see _right now_ that I'd love to get design input on. This is by no means a complete list and I expect your version of this list to be longer than that :)

I think the general idea of the reference panel - a single panel that stays open, with multiple lists that are filterable and that open links in a little side code blob that can then be expanded - is something we should stay with. It's pretty close to the Google CS reference panel and IntelliJ's version and everybody I showed it to internally likes this much better than the multi-panel layout we have right now.

But there are a lot of open questions:

1. What's the layout inside that panel?
  - We do have docs/hover-information that Google CS doesn't have for example. Where do we display that? Is it also collapsible?
  - Where does the filter text input go? I put it at the top of the panel and it's nice that it's sticky when scrolling because that allows you to scroll down, but it's also a bit weird that the layout from top to bottom is "filter -> hover info -> filterable lists"
2. Do we need more controls for the lists? Collapse/uncollapse all? How do the lists behave when filtering is used?
3. We need to display the filename/repo of the code blob we load on the side, how do we best do this so it doesn't take up too much space?
4. How do we allow users to open the side code blob in the big blob so that it's easy to find and use?
5. Do we open the panel on click everywhere or do we first jump to definition and only open it once we're on the definition?
6. ...

**Comment by jjinnii** [comment]
Hey @mrnugget, sorry for getting to this late. That looks great! I'd love for the internal release + polishing and design work to be thought of as the same thing, though, rather than two events happening side-by-side. 

I can do some structured (internal and external) user testing with your version, and we can collectively monitor internal feedback. This will help us answer the questions that you've raised (and probably surface more questions), and we can incorporate these findings into a more polished solution. :) 

If you had a general idea for when an internal release version might be ready, I can start working on a user testing one-pager for us! I can share once it's far enough along, so we can make sure it's covering all the open questions that we want answered by users.

**Comment by mrnugget** [comment]
> I'd love for the internal release + polishing and design work to be thought of as the same thing, though, rather than two events happening side-by-side.

One one hand: that's kinda what I meant to say there. I don't think design is detached from that. On the other hand: there's some backend/querying stuff that needs to be done that is unrelated to design (pagination is a hack right now, we don't fall back to search-based results, etc. see my list above).



> If you had a general idea for when an internal release version might be ready, I can start working on a user testing one-pager for us!

Sounds good! I think I can merge this in the next couple of days and will post in #dev-chat to let others know about the feature flag. Or do you think we don't announce it and only have them try it through user testing?

**Comment by jjinnii** [comment]
No, please feel free to share through #dev-chat! We can monitor feedback that comes in through that or that gets sent to our slack channel. :) When I have the user testing plan, I'll just schedule it with some random internal volunteers. Is it possible for external users to follow your instructions in #30095 to set up the prototype? If possible, I'd also like to use usertesting.com for external tests.

**Comment by efritz** [comment]
@mrnugget Thanks for the update, this view is much more clear of the plan, I think.

@jjinnii I think it would make sense given the new PR description that we don't do any extensive (non-internal/casual) user testing until we've checked some of the more pressing items off of **Stage 2**. Otherwise the panel will have some obvious deficiencies/regressions that will be hard to navigate around with users.

I think it's important to feature flag it now and start to iterate on it so we get to dogfood it and not have a single massive feature branch just accumulating the work.

I think there are open questions for for designs though, and the list in the second half of [this comment](https://github.com/sourcegraph/sourcegraph/issues/30227#issuecomment-1029888585) are still big unknowns for us and pegging some of these ideas down are probably the most beneficial on our end.

**Comment by mrnugget** [comment]
> But the other issues seem less critical, and in fact, I think user testing could help us figure out _how_ we can update the prototype design to meet these gaps.

Agree. I think pagination and fallback to search-based results are critical though. And probably collapsing/uncollapsing (#30974) since hundreds of references are a really long list to scroll. Doesn't mean that collapse/uncollapse are the best solution to that. I personally think it's better than the grouped-by-repository list we have now that's not really explorable.

> For example, I believe issue #30973 - how to display a stack of definitions would be new functionality for the reference panel, and user testing could help us to understand whether it really is a deficit and if so, how we might solve for it.

Yeah, good point. That brings up the question whether we can trust our own instincts or whether we need user testing to decide what to build. Or maybe just _how_ to build something? Because [the reactions to this Slack message](https://sourcegraph.slack.com/archives/C0W2E592M/p1626947337185400) and https://github.com/sourcegraph/sourcegraph/issues/2065 tell me that others in our team (and myself, of course) see this as a deficit - what do we do when we get 10 results from usertesting saying that it's not a deficit? De-prioritize it? Or do we put our own 10 users' opinion above external opinions?

Same goes for filtering of references. Other editors/IDEs and Google CS offer this. Colleagues told me that they "need" this - do we need user testing to tell us that we need filterable references? Or, again, do we need user testing to find out _how_ to build filtering of references?

**Comment by jjinnii** [comment]
After seeing that Slack thread for context, I agree that the internal feedback is compelling, and I think we should work on this feature! I think the core tension here is that we didn't scope what constitutes a _better reference panel_ before starting development. Therefore, we don't have great documentation of the user problems we're solving and why. (e.g. thread on displaying stack of definitions, and issue #30973 isn't linked to [#2065](https://github.com/sourcegraph/sourcegraph/issues/2065)) This tracking issue is really effective at organizing engineering work, but it's hard for me as a designer to understand the rationale behind what changes we're pulling in, which is what led me to suggest user testing to validate deficit assumptions.

As a next step, I think it'd be really helpful if you could add any context or link related pre-existing issues to the issues you've created. Then to return to the original question about when user testing should happen - maybe we can agree on which of those issues are critical for a MVP prototype then start testing. I'd personally feel a lot more confident in tackling your design questions after I have some research to drive my design.  

As a second point, I would actually argue that anything we deem non-critical could be a second iteration of the improving reference panel work. This could be a conversation we start with @macraig once we have more information on the issues in this tracking issue.

**Comment by mrnugget** [comment]
> I think the core tension here is that we didn't scope what constitutes a _better reference panel_ before starting development. Therefore, we don't have great documentation of the user problems we're solving and why. (e.g. thread on displaying stack of definitions, and issue #30973 isn't linked to [#2065](https://github.com/sourcegraph/sourcegraph/issues/2065))

Yep, I now see that there possibly was or is a misunderstanding about what we're doing here. The goal (or scope for that matter) was always to build an experimental, _technical proof-of-concept_. That's what I agreed on with @efritz and @macraig.

The rationale behind that, or what I wanted to achieve with that:

1. see what is possible if we don't use `code-intel-extensions` to implement a panel
2. find out how much work it would take to implement all the features we've said we wanted (filtering, always open, no pagination, fast loading, ...)
3. if colleagues would deem it better than what what we have, which would be an argument against implementing a panel using `code-intel-extensions`.

That's why the title of this issue was "Experiment: a better code intel panel", the PR is called "Add experimental reference panel behind feature flag" and the stage that we're in in the issue up top is called "build a proof-of-concept and get internal feedback"

It _is_ still an experiment: the code is in a single file, every name has a stupid "CoolCodeIntel" prefix to it, it uses a hack to circumvent the old reference panel, there's no tests, no storybook, etc. It's a ["build one to throw away"](https://wiki.c2.com/?PlanToThrowOneAway) prototype. And I do plan to throw it away and build it proper. So everything I do is an attempt to answer the questions up there: what's possible? how much work? is it better?

That also means it's not too late to do research and design a proper panel.

> As a next step, I think it'd be really helpful if you could add any context or link related pre-existing issues to the issues you've created.

I added an explanation to each subticket.

> Then to return to the original question about when user testing should happen - maybe we can agree on which of those issues are critical for a MVP prototype then start testing. I'd personally feel a lot more confident in tackling your design questions after I have some research to drive my design.

You tell me 😄  I _personally_ don't think it makes sense to do user testing with external users if we have known deficits that a user possibly runs into. [My comment on the user testing one pager](https://docs.google.com/document/d/1qTlcqyLzmmO_MW8UfQvskNH0TLHQ84kl8g0uETrimeM/edit?disco=AAAAUHPaWSE) contains what I think are the most critical issues:

> Pagination is not implemented, filtering + pagination is not implemented, collapse/uncollapse all is not implemented - all of these play into whether it's easy to view references with "large count of results".
> The other thing to highlight is that we only use LSIF data right now, not search-based data. Meaning: it won't work on every repository, which could lead to problems in the testing.

But that doesn't mean you should be blocked here. If you don't think these are critical (you can now try it out on Sourcegraph.com and see what you run into), then go ahead. As I said [here](https://github.com/sourcegraph/sourcegraph/issues/30227#issuecomment-1029888585), I think we can work on this concurrently.

> As a second point, I would actually argue that anything we deem non-critical could be a second iteration of the improving reference panel work. This could be a conversation we start with @macraig once we have more information on the issues in this tracking issue.

I agree, but I still have some questions I need to answer. For example: how hard is to implement pagination without going through the extension API? Is it easier or not? Can we reuse the code from the code-intel-extensions to do the fallback from precise to search based (which we know we want) or do we need our own logic? If the latter is the case then: would it maybe get easier if we don't have to go through the extension API?

(Meta side-note: next time let's have these discussions in planning or before planning)

**Comment by jjinnii** [comment]
@macraig @mrnugget Following up on our [sync discussion](https://docs.google.com/document/d/1WeW4t7uI7yeAUexGyjaNLSrf8S-ygrtHAZvsnbuSK2M/edit) for a phased rollout, I've drafted a [plan for design & implementation](https://docs.google.com/document/d/1E_6hgnFlRutelU4QI8PtFfOrxrq_wjw0UMbMMtAuQ4g/edit#heading=h.zdbbjpvibnhm) that's ready for review. It's a little bit different than the suggested phases we discussed in the meeting, so please take a look and provide feedback! From my standpoint, I tried to organize the phases such that they were digestible sizes of design work.
